confd/warp/warp.go

```diff
@@ -0,0 +1,228 @@
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "failed to read key %s from etcd", key)
+	}
+
+	version := resp.Node.Value
+	resp, err = kapi.Get(context.Background(), key+"/"+version, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read version child from key %s", key)
+	}
+
+	dogu := RawData{}
+	err = json.Unmarshal([]byte(resp.Node.Value), &dogu)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshall json from etcd")
+	}
+
+	return dogu, nil
+}
+
+func unmarshalExternal(kapi client.KeysAPI, key string) (RawData, error) {
+	resp, err := kapi.Get(context.Background(), key, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read key %s from etcd", key)
+	}
+	external := RawData{}
+	err = json.Unmarshal([]byte(resp.Node.Value), &external)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshall json from etcd")
+	}
+
+	return external, nil
+}
+
+func createHref(dogu RawData) string {
+	// remove namespace
+	parts := strings.Split(dogu["Name"].(string), "/")
+	return "/" + parts[len(parts)-1]
+}
+
+func createDoguEntry(element RawData) Entry {
+	return Entry{
+		DisplayName: element["DisplayName"].(string),
+		Href:        createHref(element),
+		Title:       element["Description"].(string),
+		Target:      "self",
+	}
+}
+
+func createExternalEntry(element RawData) Entry {
+	return Entry{
+		DisplayName: element["DisplayName"].(string),
+		Href:        element["URL"].(string),
+		Title:       element["Description"].(string),
+		Target:      "external",
+	}
+}
+
+func filterByTag(dogus []RawData, tag string) []RawData {
+	filtered := []RawData{}
+	for _, raw := range dogus {
+		if raw["Tags"] != nil {
+			tags := raw["Tags"].([]interface{})
+			if tags != nil && Contains(tags, tag) {
+				filtered = append(filtered, raw)
+			}
+		}
+	}
+	return filtered
+}
+
+func (destination *Categories) insertCategories(newCategories Categories) {
+	for _, newCategory := range newCategories {
+		destination.insertCategory(newCategory)
+	}
+}
+
+func (destination *Categories) insertCategory(newCategory *Category) {
+	for _, category := range *destination {
+		if category.Title == newCategory.Title {
+			category.Entries = append(category.Entries, newCategory.Entries...)
+			return
+		}
+	}
+	*destination = append(*destination, newCategory)
+}
+
+// JSONWriter converts the data to a json
+func jsonWriter(target string, data interface{}) error {
+	bytes, err := json.Marshal(data)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal data to json")
+	}
+
+	return ioutil.WriteFile(target, bytes, 0755)
+}
+
+func execute(configuration Configuration, kapi client.KeysAPI) {
+	reader := ConfigReader{
+		kapi:          kapi,
+		configuration: configuration,
+	}
+	categories, err := reader.readFromConfig(configuration, kapi)
+	if err != nil {
+		log.Println("Error durring read", err)
+	}
+	log.Printf("all found categories: %i", categories)
+	jsonWriter(configuration.Target, categories)
```

Comment:
**[checkstyle]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
error return value not checked (jsonWriter(configuration.Target, categories))